File: ios/Sources/TaigiKeyboard/Lexicon/Models/DictionarySearchResult.swift
// Dictionary tab 用的搜尋結果模型 — 含 source 資訊與外部辭典連結。

import Foundation

/// Search result with source information for dictionary exploration (Tab 3).
// Tab 3 詞典探索的搜尋結果。除了顯示用 roman,另存純 TL 給外部辭典 URL 使用。
struct DictionarySearchResult {
    /// Sentinel id for results synthesised from the user's custom dictionary.
    // 自訂詞庫合成結果的 sentinel id。
    static let customDictMarkerId = -2

    let id: Int
    let roman: String // Display form (POJ or TL based on user setting)
    let tl: String // Raw TL from database (for external lookup URLs)
    let hanzi: String?
    let frequency: Int
    let sources: [DictionarySource]

    /// Chhoe Taigi dictionary lookup URL for this result's TL form.
    // 對應 Chhoe Taigi 辭典的查詢 URL。
    var chhoeURL: URL? {
        ExternalLookupURLBuilder.chhoeURL(forTL: tl)
    }

    /// MOE Sutian dictionary lookup URL for this result's TL form.
    // 對應教育部臺語辭典(萌典)的查詢 URL。
    var moeURL: URL? {
        ExternalLookupURLBuilder.moeURL(forTL: tl)
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/dictionary/DictionarySearchResult.kt
// region Shared-Core Candidate
// Pure logic, Kotlin stdlib only. Eligible for cross-platform extraction.
// endregion

// 中文: Tab3 字典查詢結果值型別 — 額外帶 sources(顯示哪些字典有此詞)。
// 中文: chhoeUrl()/moeUrl() 委派給 ExternalLookupURLBuilder。

package com.siansiansu.taigikeyboard.ime.dictionary

/**
 * Search result with source information for dictionary exploration.
 *
 * URL helpers (`chhoeUrl()` / `moeUrl()`) delegate to
 * [ExternalLookupURLBuilder] so that URL-formatting logic has one home.
 */
data class DictionarySearchResult(
    val id: Int,
    val roman: String, // Display form (POJ or TL based on user setting)
    val tl: String, // Raw TL from database (for Chhoe Taigi URL)
    val hanzi: String?,
    val frequency: Int,
    val sources: List<DictionarySource>,
) {
    fun chhoeUrl(): String? = ExternalLookupURLBuilder.chhoeURL(tl)

    fun moeUrl(): String? = ExternalLookupURLBuilder.moeURL(tl)
}
